rust/osm2lanes/src/transform/tags_to_lanes/separator.rs

```diff
@@ -3,111 +3,88 @@ use celes::Country;
 use super::*;
 use crate::road::Markings;
 
-#[allow(clippy::needless_collect)]
-pub fn insert_separators(
-    lanes: Vec<Lane>,
-    tags: &Tags,
-    locale: &Locale,
-    warnings: &mut RoadWarnings,
-) -> Result<Vec<Lane>, RoadError> {
-    let left_edge = lane_to_edge_separator(lanes.first().unwrap());
-    let right_edge = lane_to_edge_separator(lanes.first().unwrap());
-    let separators: Vec<Option<Lane>> = lanes
-        .windows(2)
-        .map(|window| {
-            lanes_to_separator(window.try_into().unwrap(), &lanes, tags, locale, warnings)
-        })
-        .collect();
-    // I promise this is good code, but it might need a little explanation.
-    // If there are n lanes, there will be 1 + (n-1) + 1 separators.
-    // We interleave (zip(n+1)+flat_map) the separators with the lanes, and flatten to remove the Nones.
-    let lanes: Vec<Lane> = iter::once(left_edge)
-        .chain(separators.into_iter())
-        .chain(iter::once(right_edge))
-        .zip(lanes.into_iter().map(Some).chain(iter::once(None)))
-        .flat_map(|(a, b)| [a, b])
-        .flatten()
-        .collect();
-    Ok(lanes)
+#[derive(Clone, Copy)]
+enum DirectionChange {
+    None,
+    Same,
+    Opposite,
 }
 
-/// Given a pair of lanes
+/// Given a pair of lanes, inside to outside
 /// what should the separator between them be
-fn lanes_to_separator(
-    lanes: &[Lane; 2],
-    road: &[Lane],
+pub(super) fn lanes_to_separator(
+    lanes: [&LaneBuilder; 2],
+    road: &RoadBuilder,
     tags: &Tags,
     locale: &Locale,
     warnings: &mut RoadWarnings,
 ) -> Option<Lane> {
-    match lanes {
-        [Lane::Travel {
-            designated: LaneDesignated::Foot,
-            ..
-        }, _] => Some(Lane::Separator {
-            markings: Markings::new(vec![Marking {
-                style: MarkingStyle::KerbDown,
-                color: None,
-                width: Some(Marking::DEFAULT_WIDTH),
-            }]),
-        }),
-        [_, Lane::Travel {
-            designated: LaneDesignated::Foot,
-            ..
-        }] => Some(Lane::Separator {
+    let [inside, outside] = lanes;
+    let direction_change = match [inside.direction.some(), outside.direction.some()] {
```

Comment:
I was going to say, the rest of the PR is a bit complex 😅, but this code definitely got more readable - nice

Reply:
it would not have been worth it without something getting a bit better :)